File: 1차/3주차/이수민/최적의 행렬곱셈.py
##DP...........GG
##답을 봐도 이해못하겠음.............
##포기포기포기!!!!!!!!!!!!!!!!!!!!!!!!!!!!
##아래는 GPT가 해줬습니다....

def solution(matrix_sizes):
    n = len(matrix_sizes)
    dp = [[0] * n for _ in range(n)]

    # 행렬 곱셈 크기에 대한 초기화
    for i in range(1, n):
        dp[i - 1][i] = matrix_sizes[i - 1][0] * matrix_sizes[i - 1][1] * matrix_sizes[i][1]

    # 행렬 곱셈 크기를 3개 이상일 때 계산
    for gap in range(2, n):
        for i in range(n - gap):
            j = i + gap
            dp[i][j] = float('inf')
            for k in range(i, j):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k + 1][j] + matrix_sizes[i][0] * matrix_sizes[k][1] * matrix_sizes[j][1])

    return dp[0][n - 1]

File: 1차/3주차/고주희/최적의행렬곱셈.java
import java.util.*;

class Solution {
    public int solution(int[][] matrix_sizes) {
     
        int N = matrix_sizes.length;
        
        int[][] dp = new int[N][N];
        
        for(int i = 1; i < N; i++) {
            int b = -1;
            for (int a = 0; a < N; a++) {
                b = i + a;
                
                if(b >= N) break;
                dp[a][b] = 987654321;
                for(int p = a; p < b; p++) {
                    dp[a][b] = Math.min(dp[a][b], dp[a][p] + dp[p + 1][b] + matrix_sizes[a][0] * matrix_sizes[p + 1][0] * matrix_sizes[b][1]);
                }
            }
        }
        
        return dp[0][N-1];
    }
}
